File: bench/binarytrees/binarytrees.java
/* The Computer Language Benchmarks Game
   http://shootout.alioth.debian.org/
 
   contributed by Jarkko Miettinen
   modified by Leonhard Holz
*/

public class binarytrees {

   private final static int minDepth = 4;
   private final static int threadCount = 4;
   private final static TreeGenerator[] threads = new TreeGenerator[threadCount];
   
   public static void main(String[] args)
   {
      int n = 0;
      if (args.length > 0) n = Integer.parseInt(args[0]);
      
      int maxDepth = (minDepth + 2 > n) ? minDepth + 2 : n;
      
      TreeGenerator stretcher = new TreeGenerator(0, 0, maxDepth + 1);
      try {
         stretcher.start();
         stretcher.join();
      } catch (InterruptedException e) {
         e.printStackTrace();
      }
      int check = stretcher.getCheckResult();
      System.out.println("stretch tree of depth " + (maxDepth + 1) + "\t check: " + check);

      TreeGenerator longLived = new TreeGenerator(0, 0, maxDepth);
      try {
         longLived.start();
         longLived.join();
      } catch (InterruptedException e) {
         e.printStackTrace();
      }
      
      for (int depth = minDepth; depth <= maxDepth; depth+=2 ) {

         check = 0;
         int iterations = 1 << (maxDepth - depth + minDepth);
         int length = iterations / threadCount;

         for (int i = 0; i < threadCount; i++) {
            threads[i] = new TreeGenerator(i * length, (i + 1) * length, depth);
            threads[i].start();
         }
         for (int i = 0; i < threadCount; i++) try {
            threads[i].join();
            check += threads[i].getCheckResult();
         } catch (InterruptedException e) {
            e.printStackTrace();
         }

         System.out.println((iterations * 2) + "\t trees of depth " + depth + "\t check: " + check);
      }   

      System.out.println("long lived tree of depth " + maxDepth + "\t check: "+ longLived.getCheckResult());
   }

   private static class TreeGenerator extends Thread
   {
      private int start;
      private int end;
      private int depth;
      private int result = 0;
      
      private TreeGenerator(int start, int end, int depth)
      {
         this.end = end;
         this.start = start;
         this.depth = depth;
      }
      
      private int getCheckResult()
      {
         return result;
      }
      
      private TreeNode bottomUpTree(int item, int depth)
      {
         TreeNode node = new TreeNode();
         node.item = item;
         if (depth > 0) {
            node.left = bottomUpTree(2 * item - 1, depth - 1);
            node.right = bottomUpTree(2 * item, depth - 1);
         } else {
            node.left = null;
         }
         return node;
      }

      public void run()
      {
         if (start == end) {
            result += bottomUpTree(start, depth).itemCheck();
         } else for (int i = start; i < end; i++) {
            result += bottomUpTree(i, depth).itemCheck() + bottomUpTree(-i, depth).itemCheck();
         }
      }
   }
   
   private static class TreeNode
   {
      private int item;
      private TreeNode left, right;
      
      private int itemCheck()
      {
         if (left == null) {
            return item;
         } else {
            return item + left.itemCheck() - right.itemCheck();
         }
      }
   }
}

File: bench/binarytrees/binarytrees.go
/* The Computer Language Benchmarks Game
 * http://shootout.alioth.debian.org/
 *
 * contributed by The Go Authors.
 * based on C program by Kevin Carson
 * flag.Arg hack by Isaac Gouy
 */

package main

import (
   "flag"
   "fmt"
   "strconv"
)

var n = 0

type Node struct {
     item   int
     left, right   *Node
}

func  bottomUpTree(item, depth int) *Node {
   if depth <= 0 {
      return &Node{item: item}
   }
   return &Node{ item, bottomUpTree(2*item-1, depth-1), bottomUpTree(2*item, depth-1) }
}

func (n *Node) itemCheck() int {
   if n.left == nil {
      return n.item
   }
   return n.item + n.left.itemCheck() - n.right.itemCheck()
}

const minDepth = 4

func main() {
   flag.Parse()
   if flag.NArg() > 0 { n,_ = strconv.Atoi( flag.Arg(0) ) }

   maxDepth := n
   if minDepth + 2 > n {
      maxDepth = minDepth + 2
   }
   stretchDepth := maxDepth + 1

   check := bottomUpTree(0, stretchDepth).itemCheck()
   fmt.Printf("stretch tree of depth %d\t check: %d\n", stretchDepth, check)

   longLivedTree := bottomUpTree(0, maxDepth)

   for depth := minDepth; depth <= maxDepth; depth+=2 {
      iterations := 1 << uint(maxDepth - depth + minDepth)
      check = 0

      for i := 1; i <= iterations; i++ {
         check += bottomUpTree(i,depth).itemCheck()
         check += bottomUpTree(-i,depth).itemCheck()
      }
      fmt.Printf("%d\t trees of depth %d\t check: %d\n", iterations*2, depth, check)
   }
   fmt.Printf("long lived tree of depth %d\t check: %d\n", maxDepth, longLivedTree.itemCheck())
}
